grpc_adapter.go

```diff
@@ -0,0 +1,668 @@
+	request.ProtoMajor = 2
+	request.ProtoMinor = 0
+	header := request.Header
+	ensureTeTrailers(header)
+
+	delHeaderCanonical(header, connectHeaderProtocolVersion)
+	delHeaderCanonical(header, headerContentLength)
+
+	// Translate timeout header
+	if timeout := getHeaderCanonical(header, connectHeaderTimeout); len(timeout) > 0 {
+		delHeaderCanonical(header, connectHeaderTimeout)
+		setHeaderCanonical(header, grpcHeaderTimeout, timeout+"m")
+	}
+
+	// stream
+	if strings.HasPrefix(typ, connectStreamingContentTypeDefault) {
+		contentType := grpcContentTypePrefix + enc
+		setHeaderCanonical(header, headerContentType, contentType)
+		if contentEncoding := getHeaderCanonical(header, connectStreamingHeaderCompression); len(contentEncoding) > 0 {
+			delHeaderCanonical(header, connectStreamingHeaderCompression)
+			setHeaderCanonical(header, grpcHeaderCompression, contentEncoding)
+		}
+		if acceptEncoding := getHeaderCanonical(header, connectStreamingHeaderAcceptCompression); len(acceptEncoding) > 0 {
+			delHeaderCanonical(header, connectStreamingHeaderAcceptCompression)
+			setHeaderCanonical(header, grpcHeaderAcceptCompression, acceptEncoding)
+		}
+		return
+	}
+
+	// unary
+	contentType := grpcContentTypePrefix + strings.TrimPrefix(typ, "application/")
+	setHeaderCanonical(header, headerContentType, contentType)
+
+	if acceptEncoding := getHeaderCanonical(header, connectUnaryHeaderAcceptCompression); len(acceptEncoding) > 0 {
+		delHeaderCanonical(header, connectUnaryHeaderAcceptCompression)
+		setHeaderCanonical(header, grpcHeaderAcceptCompression, acceptEncoding)
+	}
+
+	generateGetBody := func() {
+		request.Method = http.MethodPost
+		query := request.URL.Query()
+		request.URL.RawQuery = "" // clear query parameters
+
+		compressed := false
+		if contentEncoding := query.Get(connectUnaryCompressionQueryParameter); len(contentEncoding) > 0 {
+			compressed = contentEncoding != "identity"
+			delHeaderCanonical(header, connectUnaryHeaderCompression)
+			setHeaderCanonical(header, grpcHeaderCompression, contentEncoding)
+		}
+
+		msg := query.Get(connectUnaryMessageQueryParameter)
+		msgReader := queryValueReader(msg, query.Get(connectUnaryBase64QueryParameter) == "1")
+		request.Body = &bufferedEnvelopeReader{
+			ReadCloser: io.NopCloser(msgReader),
+			compressed: compressed,
+			config:     config,
+		}
+	}
+	generatePostBody := func() {
+		compressed := false
+		if contentEncoding := getHeaderCanonical(header, connectUnaryHeaderCompression); len(contentEncoding) > 0 {
+			compressed = contentEncoding != "identity"
+			delHeaderCanonical(header, connectUnaryHeaderCompression)
+			setHeaderCanonical(header, grpcHeaderCompression, contentEncoding)
+		}
+		request.Body = &bufferedEnvelopeReader{
+			ReadCloser: request.Body,
+			compressed: compressed,
+			config:     config,
+		}
+	}
+
+	// Handle GET request with query parameters
+	if request.Method == http.MethodGet {
+		generateGetBody()
+	} else {
+		generatePostBody()
+	}
```

Comment:
Why define these as lambdas instead of just inlining here? Or maybe move GET to its own top-level named function (and pass it params instead of closing over params) so you can do something more like:
```go
if request.Method == http.MethodGet {
    translateConnectGetRequestToGRPC(...)
    return
}
// inline handling of POST here
```

Reply:
Was a linting issues, should have split out sooner.